scripts/affiliate-server

```diff
@@ -0,0 +1,211 @@
+
+start affiliate-server as fastcgi:
+
+    ./scripts/affiliate-server openlibrary.yml fastcgi 31337
+
+start affiliate-server using gunicorn webserver:
+
+    ./scripts/affiliate-server openlibrary.yml --gunicorn -b 0.0.0.0:31337
+
+"""
+
+import logging
+import os
+import sys
+import web
+import yaml
+
+from openlibrary.core import cache
+from openlibrary.core.vendors import AmazonAPI
+from openlibrary.utils.isbn import (
+    normalize_isbn, isbn_13_to_isbn_10, isbn_10_to_isbn_13)
+from openlibrary.utils.dateutil import DAY_SECS
+import _init_path
+
+logger = logging.getLogger("affiliate-server")
+
+urls = (
+    '/isbn/([0-9X-]+)', 'Submit'
+)
+
+class Submit:
+
+    @classmethod
+    def unpack_isbn(cls, isbn):
+        isbn = normalize_isbn(isbn)
+        isbn10 = isbn if len(isbn) == 10 else isbn.startswith('978') and isbn_13_to_isbn_10(isbn)
+        isbn13 = isbn if len(isbn) == 13 else isbn_10_to_isbn_13(isbn)
+        return isbn10, isbn13
+
+
+    def GET(self, isbn, threshold=4):
+        """
+        If the threshold is met, piggy back request and do an amazon lookup to fill cache.
+        Next, check if solution is in cache. If so, return it.
+        Otherwise, return the equivalent of a promise e.g. "submitted"
+        """
+
+        if not web.amazon_api:
+            return {"error": "not_configured"}
+
+        isbn10, isbn13 = self.unpack_isbn(isbn)
+        if not (isbn10 and isbn13):
+            return {
+                "error": "rejected_isbn",
+                "isbn10": isbn10,
+                "isbn13": isbn13
+            }
+
+        # Cache lookup by isbn13. If there's a miss, add to queue
+        if not cache.memcache_cache.get('amazon_product_%s' % isbn13):
+            # Misses (which will get submitted to amz as ASIN) are isbn10
+            web.amazon_queue.add(isbn10)
+
+        if len(web.amazon_queue) > threshold:
+            queue = list(web.amazon_queue)
+            web.amazon_queue = set(queue[10:])
+            #try:
+            products = web.amazon_api.get_products(queue[:10], serialize=True)
+            for product in products:
+                # Make sure we have an isbn13 for cache key
+                cache_key = (
+                    product.get('isbn_13') and product.get('isbn_13')[0] or 
+                    isbn_10_to_isbn_13(product.get('isbn_10')[0])
+                )
+                cache.memcache_cache.set(
+                    'amazon_product_%s' % cache_key, product, expires=DAY_SECS)
+            #except Exception:
+            #    logger.exception("amazon_api.get_product({}, serialize=True)".format(queue))                
+            #    web.amazon_queue.update(queue)  # Careful, request may enter fail-loop from bad IDs
+
+        if cache.memcache_cache.get('amazon_product_%s' % isbn13):
+            return {
+                "status": "success",
+                "hit": cache.memcache_cache.get('amazon_product_%s' % isbn13)
+            }
+        return {
```

Comment:
XXX blocker: All return types need to be application/json and json.dumps. Currently, it's returning a python object, not valid json

Reply:
`status` is a str, `queue` is an int, `threshold` is an int.